runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,728 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      }
+   else
+      {
+      // If unresolved at compile time, we Initialize the offset to -1. The call to generateFillInDataBlockSequenceForUnresolvedField will
+      // generate instructions to resolve and correctly populate this field at runtime.
+      instanceFieldDataBlock.offset = (UDATA) -1;
+      }
+
+   TR::S390WritableDataSnippet *dataSnippet = (TR::S390WritableDataSnippet*)cg->CreateConstant(node, &instanceFieldDataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true);
+
+   if (!isResolved)
+      {
+      // Resolve and populate dataSnippet fields.
+      generateFillInDataBlockSequenceForUnresolvedField(node, owningMethod, dataSnippet, isWrite, cg);
+      }
+
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *j9classReg = cg->allocateRegister();
+   TR::Register *j9classFlags = cg->allocateRegister();
+ 
+   // Load the class of the instance object into j9classReg.
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, j9classReg, generateS390MemoryReference(objReg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   TR_ASSERT_FATAL(J9ClassHasWatchedFields >= SHRT_MIN  && J9ClassHasWatchedFields <= SHRT_MAX, "expect value of J9ClassHasWatchedFields to be with in 16 bits");
+
+   // Test the bit in fieldClassReg with the J9ClassHasWatchedFields constant to make sure fieldwatch is enabled.
+   auto instr = generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, j9classFlags, generateS390MemoryReference(j9classReg, ((TR_J9VMBase *)(cg->fe()))->getOffsetOfClassFlags(), cg));
+   generateRIInstruction(cg, TR::InstOpCode::NILL, node, j9classFlags, J9ClassHasWatchedFields);
+   cg->setImplicitExceptionPoint(instr);
+   instr->setNeedsGCMap(0x0000FFFF);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+   // If fieldwatch is not enabled, we can branch to cFlowRegionEnd directly.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, cFlowRegionEnd);
+
+   uint8_t numOfConditions = getNumOfConditionsForReportFieldAccess(node, isResolved, isWrite, true /*isInstanceField*/, cg);
+
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions(0, numOfConditions, cg);
+   deps->addPostCondition(j9classReg, TR::RealRegister::AssignAny);
+
+   // Now we finally generate instructions to report the fieldwatch to the appropriate VM helper routine.
+   generateReportFieldAccessOutlinedInstructions(node, cFlowRegionEnd, dataSnippet, isWrite, deps, cg);
```

Comment:
Yeah the naming here is poor. On x86 they do generate some instructions OOL. IIRC, we discussed doing this on Z as well, but decided it's probably not needed because performance isn't supposed to be a big concern when fieldwatch is enabled. Do you prefer now that we also generate OOL instructions?

Reply:
Yes, performance is not too big of a concern in this case IMO. However implementation wise using a proper OOL you will no longer need an ICF region for the caller of this function. As such it simplifies the code quite a bit because you don't have to pass dependencies around.

Reply:
>However implementation wise using a proper OOL you will no longer need an ICF region for the caller of this function
I think I'm missing something here. Would you be able to explain why it would no longer be an ICF? Even if we generated those instructions inside an OOL, we still need to make a decision (using the `NILL`+`BRC` above) whether we should branch to the OOL or not. Doesn't that imply we will be in an ICF region when trying to make that decision?

Reply:
Local RA is aware of OOL code sections and it knows how to deal with the control flow. Therefore _internal_ control flow does not exist. Internal means RA is not aware of it.

If you place the contents of `generateReportFieldAccessOutlinedInstructions` in an OOL section then you no longer need the dependencies and ICF labels (`cFlowRegionStart` and `cFlowRegionEnd`) in the two places which call this function. This is all handled automatically. 

I've written some details of the differences between ICF and OOL in [1].

[1] https://github.com/eclipse/omr/blob/9b73e2bd01b72dc7ac78fac147504042396e74cb/compiler/codegen/OutOfLineCodeSection.hpp#L36-L106

Reply:
Makes sense. I've moved all VM helper calls to OOL sections: c070a0b